### PR TITLE
[WIP] Add set_nonblocking method on Stdin

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
We have `set_nonblocking` method on sockets, but not on stdin, which is very useful. So I guess it might makes sense to have it in the stdlib as well. What do you think?

### CODE REVIEW

File: src/libstd/sys/windows/stdio.rs [hunk 1]
```diff
@@ -134,6 +134,28 @@ impl Stdin {
...
+        let handle = get_handle(c::STD_INPUT_HANDLE)?;
+        let mut mode = 0;
+        if !c::GetConsoleMode(handle, &mut mode) {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "Cannot get console mode for the given handle",
+            ));
+        }
+        if nonblocking {
+            mode ^= c::ENABLE_LINE_INPUT;
```
Comment (retep998) [inline src/libstd/sys/windows/stdio.rs:147]: Won't doing xor cause `ENABLE_LINE_INPUT` to be set if it wasn't already set?

File: src/libstd/sys/unix/stdio.rs [hunk 2]
```diff
@@ -1,7 +1,10 @@
 use crate::io::{self, IoSlice, IoSliceMut};
 use crate::mem::ManuallyDrop;
+use crate::sys::cvt;
 use crate::sys::fd::FileDesc;
 
+use libc;
```
Comment (lzutao) [inline src/libstd/sys/unix/stdio.rs:6]:
```suggestion
```

File: src/libstd/sys/windows/stdio.rs (hunk 1 above)
Comment (GuillaumeGomez) [reply src/libstd/sys/windows/stdio.rs:147]: Good point.

Comment (retep998) [reply src/libstd/sys/windows/stdio.rs:148]: Perhaps it might be clearer to do `mode &= !c::ENABLE_LINE_INPUT;`?

Comment (GuillaumeGomez) [reply src/libstd/sys/windows/stdio.rs:148]: Not sure it would work as expected though...

Comment (retep998) [reply src/libstd/sys/windows/stdio.rs:148]: Why not?

Comment (GuillaumeGomez) [reply src/libstd/sys/windows/stdio.rs:148]: You're absolutely right, my bad... For further reference, I tested using [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e84c215a091145526938873fa5c793a3) to confirm it works.

**Comment by retep998** [comment]
How exactly does `O_NONBLOCK` behave for terminals on unixy platforms? On Windows when `ENABLE_LINE_INPUT` is disabled it will still block until some input is received, but it will not wait for a whole line to be entered but rather return immediately as soon as there is any input. Do unixy platforms behave the same or can they return immediately if there is no input? If the latter, then this difference needs to be spelled out big and loud in the documentation.

**Comment by GuillaumeGomez** [comment]
@retep998 I'm actually surprised that windows still blocks... When you set stdin as non blocking, it doesn't block on it, even if there is no input available. You can test it with this code:

```C
#include <unistd.h>
#include <fcntl.h>
#include <stdio.h>
#include <string.h>

int main() {
  int flags = fcntl(0, F_GETFL, 0);
  fcntl(0, F_SETFL, flags | O_NONBLOCK);
  char c[5] = {0, 0, 0, 0, 0};
  char *msg = "Enter characters: ";
  write(1, msg, strlen(msg));
  while (1) {
    int r = read(0, &c, 5);
    if (r > 0) {
      printf("Read %d characters\n", r);
      write(1, msg, strlen(msg));
    }
  }
  return 0;
}
```

**Comment by retep998** [comment]
> You can test it with this code:

I could if I had a unixy environment handy but I don't which is why I asked.

According to MSDN:

> If this mode is disabled, the functions return when one or more characters are available.

It will block if there is no input. It might be better to split this into two platform specific functions due to the significant difference in behavior.

**Comment by Mark-Simulacrum** [comment]
It looks like this isn't yet ready for review -- I don't have the bandwidth to figure out details here, so marking as S-waiting-on-author for the time being.

**Comment by GuillaumeGomez** [comment]
@retep998 What do you suggest then for the naming?

**Comment by retep998** [comment]
According to MSDN

> A process can specify a console input buffer handle in one of the wait functions to determine when there is unread console input. When the input buffer is not empty, the state of a console input buffer handle is signaled.

When stdin is a console input buffer (not a pipe or anything else), you can check whether there is input to be read, allowing for you to create a read wrapper that does not block. However, note that this specifically applies to any event that can be read using `ReadConsoleInput`, including mouse input which would not be read using normal `ReadConsoleW` calls, so the handle might be signaled but `ReadConsoleW` would still block. Also if the console is in line input mode, `ReadConsoleW` will block until the user presses enter which you have no way of knowing in advance.

If stdin is a pipe, you can set the state of the pipe to `PIPE_NOWAIT` which causes reads to be non-blocking, so if there is no input available it will return immediately. If stdin is a file, there is no non-blocking read, instead there is only overlapped reads where you start a read, it returns immediately, and then you get the result later *somehow*.

**Comment by Mark-Simulacrum** [comment]
I am going to go ahead and close this PR as it doesn't seem to be making progress, and it might be better to hash out viability elsewhere (e.g., internals) since it seems like there's some unexpected trouble with platform compatibility.

**Pull request closed without merging**